### PR TITLE
support torchrun and optimize gpu assignment

### PR DESCRIPTION
**Pull request by lxning**
## Description

Please read our [CONTRIBUTING.md](https://github.com/pytorch/serve/blob/master/CONTRIBUTING.md) prior to creating your first pull request.

Please include a summary of the feature or issue being fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

- support torchrun parameters in model config yaml file.
- support gpu assignment "deviceIds" in model config yaml file.
- update doc about model config and gpu assignment algorithm.
- set torchrun log-dir under logs

Fixes #(issue)
#2207 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- test gpu assignment
```
cat model-store/t/model-config.yaml
minWorkers: 1
maxWorkers: 1
maxBatchDelay: 100
responseTimeout: 120
parallelType: "pp"
deviceType: "gpu"
deviceIds: [2,3]

torchrun:
    nproc-per-node: 2

pippy:
    rpc_timeout: 1800
```
1. model-config.yaml
2. log
```
2023-03-30T23:45:17,829 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - [PID]67915
2023-03-30T23:45:17,830 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Torch worker started.
2023-03-30T23:45:17,830 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Python runtime: 3.8.16
2023-03-30T23:45:17,830 [DEBUG] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - W-29500-bloom_1.0 State change null -> WORKER_STARTED
2023-03-30T23:45:17,830 [DEBUG] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - W-29500-bloom_1.0 State change null -> WORKER_STARTED
2023-03-30T23:45:17,834 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /tmp/.ts.sock.29500
2023-03-30T23:45:17,834 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /tmp/.ts.sock.29500
2023-03-30T23:45:17,840 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Connection accepted: /tmp/.ts.sock.29500.
2023-03-30T23:45:17,840 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /tmp/.ts.sock.29501
2023-03-30T23:45:17,840 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /tmp/.ts.sock.29501
2023-03-30T23:45:17,842 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Connection accepted: /tmp/.ts.sock.29501.
2023-03-30T23:45:17,843 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Flushing req.cmd LOAD to backend at: 1680219917843
2023-03-30T23:45:17,843 [INFO ] W-29500-bloom_1.0 org.pytorch.serve.wlm.WorkerThread - Flushing req.cmd LOAD to backend at: 1680219917843
2023-03-30T23:45:17,858 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - model_name: bloom, batchSize: 1
2023-03-30T23:45:17,864 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - model_name: bloom, batchSize: 1
2023-03-30T23:45:18,439 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Transformers version 4.27.4
2023-03-30T23:45:18,440 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - CUDA_VISIBLE_DEVICES=2,3
2023-03-30T23:45:18,440 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - Transformers version 4.27.4
2023-03-30T23:45:18,440 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - CUDA_VISIBLE_DEVICES=2,3
2023-03-30T23:45:18,470 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - n_devs=2
2023-03-30T23:45:18,471 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - worker0, 1: 0
2023-03-30T23:45:18,471 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - worker1, 1: 1
2023-03-30T23:45:18,471 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - rank = 1 pid/device = 67916/cuda:1
2023-03-30T23:45:18,472 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - n_devs=2
2023-03-30T23:45:18,472 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - worker0, 0: 0
2023-03-30T23:45:18,472 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - worker1, 0: 1
2023-03-30T23:45:18,473 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - rank = 0 pid/device = 67915/cuda:0
2023-03-30T23:45:18,859 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - rpc.timeout=1800
2023-03-30T23:45:18,859 [INFO ] W-29500-bloom_1.0-stdout MODEL_LOG - rpc.timeout=1800
```

## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: frontend/archive/src/test/resources/modelConfig/valid.yaml [hunk 1]
```diff
@@ -5,6 +5,9 @@ batchSize: 1
 maxBatchDelay: 100
 responseTimeout: 120
 deviceType: "gpu" # cpu, gpu
-deviceIds: [0,1,2,3] # device index for gpu, neuron
-parallelLevel: 4 # rpc world size
-parallelType: "pp" # pp: pipeline parallel; pptp: tensor+pipeline parallel
\ No newline at end of file
+deviceIds: [0,1,2,3] # device index for gpu
```
Comment (HamidShojanazeri) [inline frontend/archive/src/test/resources/modelConfig/valid.yaml:8]: @lxning are we still using it?

File: docs/configuration.md [hunk 2]
```diff
@@ -266,11 +266,13 @@ models={\
...
+Starting from version 0.8.0, TorchServe allows for model configuration using a YAML file embedded in the MAR file. This YAML file contains two distinct parts that determine how a model is configured: frontend parameters and backend parameters. (see [details](https://github.com/pytorch/serve/tree/master/model-archiver#config-file))
 
 * The frontend parameters are controlled by TorchServe's frontend and specify the parameter name and default values. TorchServe now uses a priority order to determine the final value of a model's parameters in frontend. Specifically, the config.property file has the lowest priority, followed by the model configuration YAML file, and finally, the REST or gRPC model management API has the highest priority.
 
-* The backend parameters are fully controlled by the user. Users customized handler can access the backend parameters via the `model_yaml_config` property of the context object.
+* The backend parameters are fully controlled by the user. Users customized handler can access the backend parameters via the `model_yaml_config` property of the [context object](https://github.com/pytorch/serve/blob/master/ts/context.py#L24).
+
```
Comment (HamidShojanazeri) [inline docs/configuration.md:274]: adding an example how model_yaml_config can be useful + adding one how to access it in the handler.

**Review by HamidShojanazeri — APPROVED** [review]
thanks @lxning added two minor comments

**Review by msaroufim — APPROVED** [review]

File: frontend/archive/src/test/resources/modelConfig/valid.yaml (hunk 1 above)
Comment (lxning) [reply frontend/archive/src/test/resources/modelConfig/valid.yaml:8]: yes, this is unit test

**Pull request merged**